Foundation/Network/IPv6.hs

```diff
@@ -0,0 +1,157 @@
+
+-- | serialise to human readable IPv6
+--
+-- >>> toString (fromString "0:0:0:0:0:0:0:1" :: IPv6)
+toString :: IPv6 -> String
+toString = fromList . toLString
+
+toLString :: IPv6 -> [Char]
+toLString ipv4 =
+    let (i1,i2,i3,i4,i5,i6,i7,i8) = toTuple ipv4
+     in intercalate ":" $ showHex4 <$> [i1,i2,i3,i4,i5,i6,i7,i8]
+
+showHex4 :: Word16 -> [Char]
+showHex4 c
+  | c < 0x10   = '0':'0':'0':showHex c
+  | c < 0x100  = '0':'0':showHex c
+  | c < 0x1000 = '0':showHex c
+  | otherwise  = showHex c
+
+showHex :: Word16 -> [Char]
+showHex = Base.printf "%03x"
+
+fromLString :: [Char] -> IPv6
+fromLString s =
+    case parseInt' (parseHex $ snd $ span isSeparator s) [] of
+        Left err        -> error err
+        Right (addr, _) -> addr
+    where
+        parseInt' :: Either [Char] (Word16, [Char])
+                  -> [Word16]
+                  -> Either [Char] (IPv6, [Char])
+        parseInt' (Left err)           _          = Left err
+        parseInt' (Right (w8, xs))     [w7,w6,w5,w4,w3,w2,w1] = Right (fromTuple (w1, w2, w3, w4, w5, w6, w7, w8), xs)
+        parseInt' (Right (_,  []))     _          = Left "Not an ipv6 addr"
+        parseInt' (Right (w,  ':':xs)) acc        = parseInt' (parseHex xs) (w:acc)
+        parseInt' (Right (_,    c:_ )) _          = Left $ "Not an ipv6 addr: unexpected char '" <> [c] <> "'"
+
+        parseHex :: [Char]
+                 -> Either [Char] (Word16, [Char])
+        parseHex buf =
+            case span isHexDigit buf of
+                ([], x:xs) -> case x of
+                                ':' -> Right (0, x:xs)
+                                _   -> Left $ "Not an ipv6 addr: unexpected char '" <> [x] <> "'"
+                (l , xs)   -> let lhs = readHex l :: [(Word16, [Char])]
+                              in  case lhs of
+                                    [(w, [])] -> Right (w, xs)
+                                    _ -> Left "can't fall here"
+
+
+-- | create an IPv6 from the given tuple
+fromTuple :: (Word16, Word16, Word16, Word16, Word16, Word16, Word16, Word16)
+          -> IPv6
+fromTuple (i1, i2, i3, i4, i5, i6, i7, i8) = IPv6 hi low
+  where
+    f :: Word16 -> Word64
+    f = fromIntegral
+    hi, low :: Word64
+    hi =    (f i1 .<<. 48) .&. 0xFFFF000000000000
+        .|. (f i2 .<<. 32) .&. 0x0000FFFF00000000
+        .|. (f i3 .<<. 16) .&. 0x00000000FFFF0000
+        .|. (f i4        ) .&. 0x000000000000FFFF
+    low =   (f i5 .<<. 48) .&. 0xFFFF000000000000
+        .|. (f i6 .<<. 32) .&. 0x0000FFFF00000000
+        .|. (f i7 .<<. 16) .&. 0x00000000FFFF0000
+        .|. (f i8        ) .&. 0x000000000000FFFF
+
+-- | decompose an IPv6 into a tuple
+toTuple :: IPv6 -> (Word16,Word16,Word16,Word16,Word16,Word16,Word16,Word16)
+toTuple (IPv6 hi low) =
+    (f w1, f w2, f w3, f w4, f w5, f w6, f w7, f w8)
+  where
+    f :: Word64 -> Word16
+    f = fromIntegral
+    w1, w2, w3, w4, w5, w6, w7, w8 :: Word64
+    w1 = hi  .>>. 48 .&. 0x0000FFFF
```

Comment:
fromIntegral is guarantee to mask to `Word16` (`0x0000FFFF`), so this is redundant